lms/djangoapps/courseware/courses.py

```diff
@@ -137,26 +147,69 @@ def get_course_overview_with_access(user, action, course_key, check_if_enrolled=
         course_overview = CourseOverview.get_from_id(course_key)
     except CourseOverview.DoesNotExist:
         raise Http404("Course not found.")
-    check_course_access(course_overview, user, action, check_if_enrolled)
+    check_course_access_with_redirect(course_overview, user, action, check_if_enrolled)
     return course_overview
 
 
-def check_course_access(course, user, action, check_if_enrolled=False, check_survey_complete=True):
+def check_course_access(course, user, action, check_if_enrolled=False, check_survey_complete=True, check_if_authenticated=False):
     """
     Check that the user has the access to perform the specified action
     on the course (CourseDescriptor|CourseOverview).
 
     check_if_enrolled: If true, additionally verifies that the user is enrolled.
     check_survey_complete: If true, additionally verifies that the user has completed the survey.
     """
-    # Allow staff full access to the course even if not enrolled
-    if has_access(user, 'staff', course.id):
-        return
+    def _check_nonstaff_access():
```

Comment:
This private function allows us to be more efficient while keeping the code fairly readable.  We first check all the normal conditions assuming the user is not staff.  Then, if they still don't have access, only then do we bother to check if they have staff access.  See `can_load` in access.py for an example of a more verbose/repetitive version of a similar operation.